cdc/owner/changefeed.go

```diff
@@ -0,0 +1,438 @@
+	ctx = cdcContext.WithErrorHandler(ctx, func(err error) error {
+		c.errCh <- errors.Trace(err)
+		return nil
+	})
+	if err := c.tick(ctx, state, captures); err != nil {
+		log.Error("an error occurred in Owner", zap.String("changefeedID", c.state.ID), zap.Error(err), zap.Stringer("tp", reflect.TypeOf(err)))
+		var code string
+		if rfcCode, ok := cerror.RFCCode(err); ok {
+			code = string(rfcCode)
+		} else {
+			code = string(cerror.ErrOwnerUnknown.RFCCode())
+		}
+		c.feedStateManager.HandleError(&model.RunningError{
+			Addr:    util.CaptureAddrFromCtx(ctx),
+			Code:    code,
+			Message: err.Error(),
+		})
+		c.releaseResources()
+	}
+}
+
+func (c *changefeed) tick(ctx cdcContext.Context, state *model.ChangefeedReactorState, captures map[model.CaptureID]*model.CaptureInfo) error {
+	c.state = state
+	c.feedStateManager.Tick(state)
+	if !c.feedStateManager.ShouldRunning() {
+		c.releaseResources()
+		return nil
+	}
+
+	checkpointTs := c.state.Info.GetCheckpointTs(c.state.Status)
+	if err := c.gcManager.CheckStaleCheckpointTs(ctx, checkpointTs); err != nil {
+		return errors.Trace(err)
+	}
+	if !c.preflightCheck(captures) {
+		return nil
+	}
+	if err := c.initialize(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	select {
+	case err := <-c.errCh:
+		return errors.Trace(err)
+	default:
+	}
+
+	c.sink.EmitCheckpointTs(ctx, checkpointTs)
+	barrierTs, err := c.handleBarrier(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	shouldUpdateState, err := c.scheduler.Tick(c.state, c.schema.AllPhysicalTables(), captures)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if shouldUpdateState {
+		c.updateStatus(barrierTs)
+	}
+	return nil
+}
+
+func (c *changefeed) initialize(ctx cdcContext.Context) error {
+	if c.initialized {
+		return nil
+	}
+	// clean the errCh
+	// When the changefeed is resumed after being stopped, the changefeed instance will be reused,
+	// So we should make sure that the errCh is empty when the changefeed is restarting
+LOOP:
+	for {
+		select {
+		case <-c.errCh:
+		default:
+			break LOOP
+		}
+	}
+	startTs := c.state.Info.GetCheckpointTs(c.state.Status)
+	log.Info("initialize changefeed", zap.String("changefeed", c.state.ID),
+		zap.Stringer("info", c.state.Info),
+		zap.Uint64("checkpoint ts", startTs))
+	failpoint.Inject("NewChangefeedNoRetryError", func() {
+		failpoint.Return(cerror.ErrStartTsBeforeGC.GenWithStackByArgs(startTs-300, startTs))
+	})
+
+	failpoint.Inject("NewChangefeedRetryError", func() {
+		failpoint.Return(errors.New("failpoint injected retriable error"))
+	})
+
+	if c.state.Info.Config.CheckGCSafePoint {
+		err := util.CheckSafetyOfStartTs(ctx, ctx.GlobalVars().PDClient, c.state.ID, startTs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	c.barriers.Update(ddlJobBarrier, startTs)
+	c.barriers.Update(syncPointBarrier, startTs)
+	c.barriers.Update(finishBarrier, c.state.Info.GetTargetTs())
```

Comment:
if `syncpoint` is not enabled or `finished ts` is not set, what about ignore these two barriers. Then we don't need to check them in `handleBarrier`